Slashing protocol documentation (#2038)
nucypher/nucypher: docs/source/architecture/slashing.rst

3. Ramped penalties, that increase with each successive violation, potentially resetting in a specified number of periods. This will encourage stakers to avoid repeat offences.
4. Temporal limitations on penalties, for example capping the total number of tokens slashabe in each period. This addresses the uneven distribution of effective punishment based on the unpredictable frequency with which a given Bob makes requests to an Ursula. Because Ursulas have no control over Bob’s behavior, they must be given a chance to rectify their worker’s incorrect re-encryptions before their stake is wiped out. This is particularly risky if a Bob is making requests at a high cadence or batches their requests. A limit on the penalty size per day addresses this unfair scenario.
5. Temporary unbonding of Ursula, which forces the staker to forfeit subsidies, work and fees for a specified period. In a simple construction, this punishment would only apply if the Ursula is not servicing any other policies or all Alices consent to the removal of that Ursula from their policies.

> 6. In conjunction with (5), delegated tokens being automatically returned to delegators. Delegators must proactively re-delegate to the offending staker after a specified time delay has elapsed.

Reviewer: I think this is not really feasible unless we would provide everything for delegating. I want to point that for slashing protocol there are no any delegators, only one staker and one worker

Author: Roger that! I'll remove these

> 7. Delegated tokens suffering the same fate as staker-owned tokens, as described in 1, 2, and 3. This strongly encourages delegators to choose the most reliable staker(s).  

Reviewer: The item is knocked out (not sure about this word) of the remaining, because it's just recommendation and not what we plan to do with slashing protocol

The Slashing Protocol
=====================

The slashing protocol is a preventative mechanism that disincentivizes certain staker actions, whether deliberate or unintentional, to maximize service quality and preserve network health. If prohibited actions (‘violations’) are attributably detected at any moment, the protocol responds by irreversibly deleting (‘slashing’) a portion of the offending staker’s collateral (‘stake’).

At network genesis, the protocol will be able to detect and attribute instances of incorrect re-encryptions returned by Ursulas. The staker controlling the incorrectly re-encrypting Ursula will have their stake reduced by a nominal number of NU tokens.

Violations
----------

> In response to an access request by Bob, Ursula must generate a ciphertext that perfectly corresponds to the associated sharing policy (i.e. precisely what Alice intended Bob to receive). If the ciphertext is invalid in this regard, then Ursula is deemed to be incorrectly re-encrypting. Each instance of incorrect re-encryption is an official violation and is individually punished.

Author: ```suggestion
In response to an access request by Bob, Ursula must generate a re-encrypted ciphertext that perfectly corresponds to the associated sharing policy (i.e. precisely what Alice intended Bob to receive). If the ciphertext is invalid in this regard, then Ursula is deemed to be incorrectly re-encrypting. Each instance of incorrect re-encryption is an official violation and is individually punished.
```

> In response to an access request by Bob, Ursula must generate a ciphertext that perfectly corresponds to the associated sharing policy (i.e. precisely what Alice intended Bob to receive). If the ciphertext is invalid in this regard, then Ursula is deemed to be incorrectly re-encrypting. Each instance of incorrect re-encryption is an official violation and is individually punished.

Author: Agree with change, just an 'a' missing ^

There are other ways stakers can compromise service quality and network health, such as extended periods of downtime or ignoring access requests. Unlike incorrect re-encryptions, actions of this kind are not yet reliably attributable. Punishing non-attributable actions may result in unacceptable outcomes or introduce perverse incentives, thus these actions are not yet defined as violations by the slashing protocol.  
Detection
----------

Incorrect re-encryptions are detectable by Bob, who can then send a proof to the protocol to confirm the violation. This is enabled by a bespoke zero-knowledge correctness verification mechanism, which follows these steps:

> 1. When Alice creates a Kfrag, it includes components to help Ursula prove the correctness of each re-encryption she performs. The Kfrag’s secret component (*bn_key*) is used to perform the re-encryption operation. The Kfrag also comprises public components, including a point commitment on the value of the bn_key.

Author: Both good points! I'll update the style and remove overly technical references

2. When Ursula receives the Kfrag, she checks it is valid – that the point commitment on the secret component is correct. This ensures that she doesn’t incorrectly re-encrypt due to Alice’s error (or attack).
3. Bob makes a re-encryption request by presenting a capsule to Ursula, and she responds with a Cfrag. This contains the payload (a re-encrypted ciphertext) and a non-interactive zero knowledge proofs of knowledge (NIZK).
4. Bob checks the validity of the Cfrag using the NIZK. He verifies that the point commitment corresponds to the ciphertext. He also checks that the Cfrag was generated using his capsule, by verifying that it was created with the correct public key.

> 5. If any of the verifications fail, then Bob supplies the ciphertext and NIZK to the Adjudicator contract. The contract runs extensive verification processes, leveraging optimized ECC algorithms.

Author: Good idea, I also added a link to Numerology

6. If the invalidity of the Cfrag is confirmed by the Adjudicator contract, the delivery of a faulty Cfrag to Bob is ruled to be an official protocol violation. A penalty is computed and the owner of the offending Ursula has their stake immediately slashed by the penalty amount.

Penalties
---------

> At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size. For each violation, 2*10^-18 NU tokens will be deleted from the offender’s stake. The maximum number of tokens that can be slashed in a given period is limited by the number of blocks processed on Ethereum (assuming 5760 per day and one slash per block) – approximately 10^-12 NU slashable per period.

Author: Haha yeh this is confusing. 2x10^-18 * 5760 is roughly 11.5^-15. But unless @vzotova got the maths wrong (in which case we should change the rules of maths, they're obviously silly), there are some other limiting factors at play here, like the gas cost per transaction. Vicky if you get a sec pls explain!

> At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size. For each violation, 2*10^-18 NU tokens will be deleted from the offender’s stake. The maximum number of tokens that can be slashed in a given period is limited by the number of blocks processed on Ethereum (assuming 5760 per day and one slash per block) – approximately 10^-12 NU slashable per period.

Reviewer: could be many slashings per block, each slashing costs different value (> 300,000 gas) and each block is 8 000 000 gas

> At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size. For each violation, 2*10^-18 NU tokens will be deleted from the offender’s stake. The maximum number of tokens that can be slashed in a given period is limited by the number of blocks processed on Ethereum (assuming 5760 per day and one slash per block) – approximately 10^-12 NU slashable per period.

Reviewer 2: @arjunhassard , after chatting with @vzotova about this, wdyt about the following:

```suggestion
At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size.
For each violation, :math:`2 \times 10 ^ {-18}` NU tokens will be deleted from the offender’s stake. The theoretical maximum number of tokens
that can be slashed in a given period is limited by the number of blocks processed on Ethereum per day (~6000) and the number of
transactions per block (~30 based on transaction size and current gas limits). This yields a maximum slashable value of:

    .. math::

        &= 2 \times 10 ^ {-18} NU \times 6000 \text{ blocks per period} \times 30 \text{ transactions per block} \\
        &= 3.6 \times 10 ^ {-13} NU \text{ per period}

```

> At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size. For each violation, 2*10^-18 NU tokens will be deleted from the offender’s stake. The maximum number of tokens that can be slashed in a given period is limited by the number of blocks processed on Ethereum (assuming 5760 per day and one slash per block) – approximately 10^-12 NU slashable per period.

Reviewer: I'd change `transaction size` to something with gas

> At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size. For each violation, 2*10^-18 NU tokens will be deleted from the offender’s stake. The maximum number of tokens that can be slashed in a given period is limited by the number of blocks processed on Ethereum (assuming 5760 per day and one slash per block) – approximately 10^-12 NU slashable per period.

Reviewer 2: ☝️ modified to `transaction gas`...?

> At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size. For each violation, 2*10^-18 NU tokens will be deleted from the offender’s stake. The maximum number of tokens that can be slashed in a given period is limited by the number of blocks processed on Ethereum (assuming 5760 per day and one slash per block) – approximately 10^-12 NU slashable per period.

Author: @derekpierre @vzotova nice one, thanks for looking into that!

4. Temporal limitations on penalties, for example capping the total number of tokens slashabe in each period. This addresses the uneven distribution of effective punishment based on the unpredictable frequency with which a given Bob makes requests to an Ursula. Because Ursulas have no control over Bob’s behavior, they must be given a chance to rectify their worker’s incorrect re-encryptions before their stake is wiped out. This is particularly risky if a Bob is making requests at a high cadence or batches their requests. A limit on the penalty size per day addresses this unfair scenario.

5. Temporary unbonding of Ursula, which forces the staker to forfeit subsidies, work and fees for a specified period. In a simple construction, this punishment would only apply if the Ursula is not servicing any other policies or all Alices consent to the removal of that Ursula from their policies.

6. In conjunction with (5), delegated tokens being automatically returned to delegators. Delegators must proactively re-delegate to the offending staker after a specified time delay has elapsed.

> 7. Delegated tokens suffering the same fate as staker-owned tokens, as described in (1), (2), and (3). This strongly encourages delegators to choose the most reliable staker(s).  

Author: It is quite specific, agreed. In addition to delegation pools, it's certainly possible for the protocol to enforce rules regarding temporary undelegation/redelegation. Indeed, for some chains its a major discincentive (see https://github.com/nucypher/nucypher/issues/1916#issuecomment-622165037). @mswilkison what do you think? i think a list of possible extensions to the slashing protocol can be fairly broad, but happy to remove!

The Slashing Protocol
=====================

> The slashing protocol is a preventative mechanism that disincentivizes certain staker actions, whether deliberate or unintentional, to maximize service quality and preserve network health. If prohibited actions (‘violations’) are attributably detected at any moment, the protocol responds by irreversibly forfeiting (‘slashing’) a portion of the offending staker’s collateral (‘stake’).

Reviewer 3: ```suggestion
The slashing protocol is a preventative mechanism that disincentivizes certain staker actions, whether deliberate or unintentional, that may negatively impact service quality or network health. If prohibited actions (‘violations’) are attributably detected at any moment, the protocol responds by irreversibly forfeiting (‘slashing’) a portion of the offending staker’s collateral (‘stake’).
```

1. When Alice creates a kFrag, it includes components to help Ursula prove the correctness of each re-encryption she performs. The kFrag’s secret component is used to perform the re-encryption operation. The kFrag also comprises public components, including a point commitment on the value of the secret component.
2. When Ursula receives the kFrag, she checks its validity – that the point commitment on the secret component is correct. This ensures that she doesn’t incorrectly re-encrypt due to Alice’s error (or attack).
3. Bob makes a re-encryption request by presenting a capsule to Ursula, and she responds with a cFrag. This contains the payload (a re-encrypted ciphertext) and a non-interactive zero knowledge proofs of knowledge (NIZK).
4. Bob checks the validity of the cFrag using the NIZK. He verifies that the point commitment corresponds to the ciphertext. He also checks that the cFrag was generated using his capsule, by verifying that it was created with the correct public key.

> 5. If any of the verifications fail, then Bob supplies the ciphertext and NIZK to the :ref:`Adjudicator contract <contracts>`. The contract runs extensive verification processes, leveraging `optimized ECC algorithms <https://github.com/nucypher/numerology>`_.

Reviewer 4: ```suggestion
5. If any of the verifications fail, then Bob supplies the ciphertext and NIZK to the :ref:`Adjudicator contract <contracts>`. The contract examines Bob's claim by checking whether the NIZK proof for the ciphertext fails, leveraging `optimized ECC algorithms <https://github.com/nucypher/numerology>`_.
```
Numerology as such is not used anymore.

6. If the invalidity of the cFrag is confirmed by the Adjudicator contract, the delivery of a faulty cFrag to Bob is ruled to be an official protocol violation. A penalty is computed and the owner of the offending Ursula has their stake immediately slashed by the penalty amount.

.. image:: ../.static/img/correctness_verification_schematic.png
    :target: ../.static/img/correctness_verification_schematic.png

Penalties
---------

At network genesis, although violations will be detected, attributed and publicly logged, the actual penalty levied will be of nominal size.
For each violation, :math:`2 \times 10 ^ {-18}` NU tokens will be deleted from the offender’s stake. The theoretical maximum number of tokens
that can be slashed in a given period is limited by the number of blocks processed on Ethereum per day (~6000) and the number of
transactions per block (~30 based on transaction gas and current gas limits). This yields a maximum slashable value of:

    .. math::

        &= 2 \times 10 ^ {-18} NU \times 6000 \text{ blocks per period} \times 30 \text{ transactions per block} \\
        &= 3.6 \times 10 ^ {-13} NU \text{ per period}

> This nominal penalty is effectively a placeholder until a more complete slashing model is designed and implemented. The genesis penalty is measurable – so staker behavior can be observed – but small enough that it has a negligible impact on the staker’s ability to continue serving the network. If the severity of penalties and logic of the slashing protocol changes, it may involve any combination of the following:

Reviewer 3: ```suggestion
The genesis penalty is measurable – so staker behavior can be observed – but small enough that it has a negligible impact on the staker’s ability to continue serving the network. If the severity of penalties and logic of the slashing protocol changes, it may involve any combination of the following:
```
